egui/src/widgets/text_edit/builder.rs

```diff
@@ -81,17 +86,15 @@ impl<'t> TextEdit<'t> {
 }
 
 impl<'t> TextEdit<'t> {
-    /// No newlines (`\n`) allowed. Pressing enter key will result in the `TextEdit` losing focus (`response.lost_focus`).
-    pub fn singleline(text: &'t mut dyn TextBuffer) -> Self {
-        Self {
-            desired_height_rows: 1,
-            multiline: false,
-            ..Self::multiline(text)
-        }
+    /// Create a new singleline `TextEdit` with the default validator.
+    pub fn new(text: &'t mut dyn TextBuffer) -> TextEdit<'t, String> {
+        Self::with_validator(text)
     }
+}
 
-    /// A `TextEdit` for multiple lines. Pressing enter key will create a new line.
-    pub fn multiline(text: &'t mut dyn TextBuffer) -> Self {
+impl<'t, V> TextEdit<'t, V> {
+    /// Set the input validator type.
+    pub fn with_validator(text: &'t mut dyn TextBuffer) -> TextEdit<'t, V> {
```

Comment:
Did you consider the alternative of passing a function pointer instead? i.e. `validator: fn(data: InputData<'_>) -> Action` or `validator: &dyn Fn(data: InputData<'_>) -> Action`. This makes it more flexible, allowing dynamic filters, and allows us to pass the validator as another builder step: `TextEdit::singleline(text).with_validator(&my_validator)`

Reply:
Interesting idea, I think its better design indeed. I'm not to used to the immediate semantics of basically creating temporary structures each frame, so I didn't even think about it.

Reply:
Also solves the naming problem.

Reply:
Yep its much nicer.